QUIC Conformance Updates: Miscellaneous

hlandau opened the pull request:
Fixes https://github.com/openssl/project/issues/36

Built on top of the shutdown drainage PR.

```
b164838bd3 QUIC: Test crypto stream FC limits
b365463e35 QUIC CONFORMANCE: Test that CRYPTO frames with bad offsets/lengths are rejected
aeaba0530c QUIC WIRE: RFC 9000 s. 19.6
0a6cf56c01 QUIC CHANNEL: Apply flow control to CRYPTO streams
17bc244a94 QUIC FC: Rename stream count mode to reflect actual function
6a33ae1f74 QUIC CHANNEL: Fix typo
999bd0e071 QUIC QRX: Test for 1-RTT processing restriction
5bda0280cf QUIC QRX: Don't process 1-RTT packets until handshake is complete
673b5d6b0b QUIC QRX: Enforce PN monotonicity with key updates
c37568f726 QUIC CHANNEL, TXP: Discard INITIAL EL correctly
58e143afb5 QUIC: Update no-TPARAM test for correct error code
39d6fa0a7d QUIC TLS: Report TLS errors properly as QUIC protocol errors
910f584b19 QUIC CHANNEL: Send correct alert code if no TPARAMs received
80bf44c8f7 QUIC TXP: Allow PATH_RESPONSE to force padding
01d9ead721 QUIC CFQ: Unreliable transmission for PATH_RESPONSE
dfd66568ef QUIC: Echo PATH_CHALLENGE frames as PATH_RESPONSE frames
```

mattcaswell commented on ssl/quic/quic_tls.c:
```diff
@@ -28,6 +29,18 @@ struct quic_tls_st {
+    uint64_t error_code;
+
+    /*
+     * Error message with static storage duration. Valid only if inerr is 1.
```
inerror

mattcaswell commented on ssl/quic/quic_tls.c:
```diff
@@ -628,6 +641,20 @@ void ossl_quic_tls_free(QUIC_TLS *qtls)
+    qtls->error_code = error_code;
+    qtls->error_msg  = error_msg;
+    qtls->inerror    = 1;
+    return 0;
```
Eventually we call ERR_raise_data when `ossl_quic_channel_raise_protocol_error` gets called....but by then we've lost the original file/line of the error. Shouldn't we be propagating the file/line of the error? E.g. see the definition of the SSLfatal macro which ensures `OPENSSL_FILE` and `OPENSSL_LINE` are captured. This probably applies everywhere else that `ossl_quic_channel_raise_protocol_error` gets called too

mattcaswell commented on test/quic_multistream_test.c:
```diff
@@ -3025,6 +3025,65 @@ static const struct script_op script_41[] = {
+
+    if (!TEST_true(WPACKET_quic_write_vlint(&wpkt, OSSL_QUIC_FRAME_TYPE_CRYPTO))
+        || !TEST_true(WPACKET_quic_write_vlint(&wpkt, (((uint64_t)1) << 62) - 1))
+        || !TEST_true(WPACKET_quic_write_vlint(&wpkt, 1))
```
I wonder if additionally checking maximum offset *and* maximum length would be worthwhile? i.e. to check no overflows etc.

hlandau replied:
This will come in the Assorted Tests PR.

hlandau commented:
Updated with better error handling (last commit).

paulidale commented on include/internal/quic_error.h:
```diff
@@ -35,7 +35,44 @@
+    QUIC_ERR_CRYPTO_ERR(TLS1_AD_NO_APPLICATION_PROTOCOL)
+
+static ossl_inline ossl_unused const char *
+ossl_quic_err_to_string(uint64_t error_code)
```
Not sure if a header is the best place for this.
It keeps it with the errors which is a plus, but it seems odd to have error handling as an inline function.

Similarly for quic_wire.h's frame_type_to_string.

hlandau commented:
@paulidale Updated.

t8m commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -889,9 +897,41 @@ static int depack_do_frame_path_challenge(PACKET *pkt,
+     * respond by echoing the data contained in the PATH_CHALLENGE frame in a
+     * PATH_RESPONSE frame.
+     *
+     * TODO(QUIC): We should try to avoid allocation here in the future.
```
Please use TODO(QUIC FUTURE)

t8m reviewed (approved):
Approved with the one comment nit fixed.

hlandau commented:
Regression during rebase, investigating.

hlandau commented:
Rebased again and fixed bug around QUIC_TLS error handling.

I've substantially rethought how to do QUIC_TLS error handling, so I recommend reviewing the last commit closely. It also adds some new APIs to libcrypto.

```
422dd29233 QUIC TLS: Rethink error handling
```

mattcaswell commented on crypto/err/err_save.c:
```diff
@@ -43,6 +43,71 @@ void OSSL_ERR_STATE_save(ERR_STATE *es)
+         thread_es->bottom != top
+         && thread_es->err_marks[top] == 0;
+         ++count)
+        top = top > 0 ? top - 1 : ERR_NUM_ERRORS - 1;
```
Seems virtually identical to ERR_count_to_mark(). Is there no opportunity for code reuse?

hlandau replied:
I might factor this in a subsequent PR. Not doing this now so we can get this in.

mattcaswell reviewed (approved):
LGTM. One minor comment below. Approved anyway if you don't do this.

mattcaswell commented:
@t8m - can you reconfirm your earlier approval?

mattcaswell commented:
Pushed. Thanks!
